### ISSUE TITLE
Error when using some dictionaries for single words resulting in SAPI TTS failure for single words

### ISSUE THREAD

**Issue by TheNicker** [goldendict/goldendict#1342]
Reproduced on several **32 bit** binary versions:  Apr 2019, Dec 17 and May 16 
System: **windows 10 64 bit version 2004**
For example: adding the dictionary Cambridge Advanced Learners Dictionary 3th Ed. (En-En) creates the following error (highlighted in red):

![image](https://user-images.githubusercontent.com/4753283/106349604-dc648d00-62d7-11eb-803b-40675d7a2f17.png)

This causes Text to speech to work on multi word idioms but not on most of the the **single words** like "hello" 
So it's a double whammy, the error from the dictionary and the subsequent failure of the TTS.

My workaround is to add a second dummy word to be able to pronounce the first one or to disable the Cambridge dictionary.

Shouldn't the dictionaries / TTS engines be decoupled ?

**Comment by vedgy** [comment]
I don't know how TTS works in GoldenDict. I wonder: is this similar to #970? In other words, is TTS triggered just like word pronunciation in the user interface - via the auto-pronounce options or by triggering the *Pronounce Word* action on the toolbar or by triggering an audio button in a specific article? If that is the case, does triggering the audio button in the TTS engine section still work, and only the auto-pronounce/toolbar button fail?

**Comment by vedgy** [comment]
A possible workaround is to create a separate dictionary group for TTS.

**Comment by TheNicker** [comment]
> is this similar to #970      

It does seems to overlap (a duplicate of ?) #970 , Does #970 gets reproduce with multiple word idioms ? 

> I don't know how TTS works in GoldenDict.

At least under windows, it uses SAPI (Speech API), so it just calls an external interface, however the problem is definitely internal.

>does triggering the audio button in the TTS engine section still work, and only the auto-pronounce/toolbar button fail?

Both seems to fail.

> A possible workaround is to create a separate dictionary group for TTS.

I wasn't aware of groups, it does seems a better (still inconvenient) workaround, but please note that I had to disable and then reenable dictionaries, for groups to take effect.

**Comment by Abs62** [comment]
File "res.zip" from your Stardict dictionaries shouldn't contain "res" folder, only files. If you have zip with "res" folder inside just extract it beside main dictionary file or repack "res.zip" without "res" folder. If you don't have "res.zip" or "res" folder then you don't have sound/image files for dictionary.

**Comment by vedgy** [comment]
> Does #970 gets reproduce with multiple word idioms ?

The bug does not depend on whether a single word or a phrase translation is requested. But rather on whether there is an audio link for the word/phrase in the top/high-priority dictionary, in which some audio resources are missing. Audio links are more common in single-word articles than in multiple-word articles, so that can explain your experience.

Yet another workaround is to reorder your dictionaries - place the TTS engine at the top. Hopefully an audio resource will be successfully requested by the name used in the TTS dictionary then.

#972 fixes #970 for me, but additional code changes may be required to fix the TTS issue, because TTS code paths are somewhat different.

**Comment by TheNicker** [comment]
@vedgy,  thanks for the clarification

@Abs62, The following image depicts the contents of the folder I'm adding to the dictionaries search path pertaining to the aforementioned dictionary:
![image](https://user-images.githubusercontent.com/4753283/106355916-f7e58d00-6303-11eb-989a-7f9b99939466.png)

It can be downloaded from: http://download.huzheng.org/bigdict/ fifth from the top

**Comment by Abs62** [comment]
The "res" folder in this dictionary doesn't contain audio files, only pictures. In this case GD tries to find file in other dictionaries in group.  Try to find same dictionary with full resource set. Its zip size should be about 300 Mb, not 60.

**Comment by TheNicker** [comment]
So if I get this right, I guess a proper solution is to first, create a TTS fallback in case the dictionary doesn't contain the **expected** (audio) resource and additionally create an additional GUI configuration per dictionary whether to use the supplied audio resource or to use TTS.

**Comment by Abs62** [comment]
Audio links in the CALD article in this case will not work anyway. Because GD tries to find required audio file, not target word. Look these links in the inspector via mouse right click. But links in TTS article should work.

**Comment by TheNicker** [comment]
Though I think it solves all cases, whether the resources are present or not. 
If it's not the optimal solution, what do you think is ?

**Comment by Abs62** [comment]
I think it is better to find dictionary with complete resources if pronunciation is needed.

**Comment by vedgy** [comment]
> create an additional GUI configuration per dictionary whether to use the supplied audio resource or to use TTS.

Why cannot a user just prioritize the TTS engine "dictionary" to the top if he/she prefers TTS over pronunciation? Doesn't this work already?

**Comment by TheNicker** [comment]
@vedgy , It does work. it helps with my use case.

There is coupling in functionality, so IMHO, it still is a workaround and not a clean solution.
After getting more input from you two, decoupling could be done by creating another "dictionary order" just for pronunciation.
And of course fall back to the next in the list upon failure.

![image](https://user-images.githubusercontent.com/4753283/106366813-18830680-6347-11eb-83a3-b16950c65ad9.png)

**Comment by vedgy** [comment]
I guess few users would want to tweak the pronunciation order. So if this feature is to be implemented, I think it should be disabled by default. For example, a checkbox *Customize pronunciation order* with a tooltip that explains that when it is disabled, the dictionary order is reused for pronunciation.

Another issue is vertical space in the Dictionaries dialog. The new feature should be both usable and most unobtrusive in this respect. Perhaps hide the *Pronunciation order* section while the customization checkbox is unchecked? The users who decide to customize the pronunciation order can resize the Dictionaries dialog then, and those who don't use the feature only pay for one checkbox UI clutter.

When the customization is enabled, there should probably be a button to copy the dictionary-order list to the pronunciation-order list for tweaking. It would be most useful after the dictionary list is changed.

Yet another consideration is that this implementation would only be useful to the users who don't use dictionary groups. In order to make it usable to more advanced users, the pronunciation order customization would have to be implemented on the Groups tab as well...

**Comment by TheNicker** [comment]
Indeed. valid points.

Another option could be creating another level of tab stacks. two under `Dictionaries` and two under `Groups`
The tabs names would be for example : `Translation` and `Pronunciation`
`Pronunciation` would be a mirror of `Pronunciation` until  `Pronunciation` is activated and the user changes priority. Then a warning pops, up for example: _"Would you like detach translation order from pronunciation order ? , to reattach click reattach from the context menu"_

* Eliminates the need for checkbox.
* Dictionaries and Groups tabs content doesn't change (unobtrusive).
* Easier to reuse existing GUI code.

**Comment by vedgy** [comment]
> Eliminates the need for checkbox.

The additional level of tab stacks would probably clutter the UI even more than the checkbox.

A downside of separate tabs is the difficulty of comparing the translation and pronunciation orders. This is most relevant when there are many (e.g. tens of) dictionaries in a group. Also if some dictionaries are excluded/removed from the pronunciation order list, how can the user move them back from the translation order group list if the two lists are on separate tabs?

Maybe the pronunciation order list should be placed to the right of the translation order list. Almost everyone has a wide screen by now.

**Comment by TheNicker** [comment]
Can you imagine a real-life scenario where there's a need to compare the two ? 

I see not, because these provide **completely separate** functionalities, although they are extracted from the same database (dictionary), but a database may provide for more then one _service_. Moreover, behavior of translation and pronunciation do not depend on one another.

**Comment by vedgy** [comment]
> Can you imagine a real-life scenario where there's a need to compare the two ?

Yes, I can. I would usually prefer to hear the sound from the same dictionary I'm reading the translation from. So my two dictionary orders would usually be pretty close to each other. Pronunciations can be different in different dictionaries, but they should match the text from their dictionary's translation. There are British English dictionaries with pronunciations and American English. Some dictionaries provide whole sentences as examples along with the pronunciations of these sentences.

I'd make exceptions (move down in the pronunciation priority list) for dictionaries with good text translations but low-quality or missing pronunciations. And maybe I'd move some dictionaries with excellent pronunciations to the top of the pronunciation priority list. But overall I am usually going to keep the two orders almost in sync.

Other than comparing the two orders, a user may want to remove some of the group's dictionaries from the pronunciation list. But later the user may decide to move them back into the pronunciation list. Rather than looking for these dictionaries in the entire dictionary list, it would be easier to drag them from the group's translation list. Keep in mind that groups can be used to separate dictionaries for different language pairs, online from offline dictionaries, in which cases it makes perfect sense to have  the same dictionaries in the two lists, but possibly in a somewhat different order.

On the other hand, I see now that a user may want to have dictionaries in a group's pronunciation list, but not in the same group's translation list. But this use case is also covered if the group pronunciation order list is on the same page with the group translation list and the entire dictionary list - the user can drag from any of these lists. Though I suspect it may be difficult to implement support for pronunciation-only dictionaries in a group.

**Comment by TheNicker** [comment]
The good is that currently we both agree on the pronunciation order concept.
Regarding the implementation, I think the use cases nuances have exhausted this medium and I got the feeling it'll get more complex from now on.

I'll be happy to continue via IM when it'll be relevant.

**Comment by vedgy** [comment]
> I'll be happy to continue via IM when it'll be relevant.

Fine with me.

@Abs62, would you be willing to review and eventually merge the pronunciation order feature in one form or another? Or do you believe this would be too much complexity for the feature unneeded by most users? If so, can you suggest an alternative simpler feature that would allow to at least select the top-priority pronunciation dictionary?

**Comment by Abs62** [comment]
>> Or do you believe this would be too much complexity for the feature unneeded by most users?

Exactly.

>> If so, can you suggest an alternative simpler feature that would allow to at least select the top-priority pronunciation dictionary?

Just place it on top of group dictionaries list.
